pkg/registry/common/nslabelsmatching/client_test.go

```diff
@@ -0,0 +1,211 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nslabelsmatching_test
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.uber.org/goleak"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/common/nslabelsmatching"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
+	"github.com/networkservicemesh/sdk/pkg/tools/sandbox"
+)
+
+const (
+	testEnvName                      = "NODE_NAME"
+	testEnvValue                     = "testValue"
+	destinationTestKey               = "nodeName"
+	destinationTestCorrectTemplate   = "{{index .Src \"NodeNameKey\"}}"
+	destinationTestIncorrectTemplate = "{{index .Src \"SomeRandomKey\"}}"
+	destinationTestNotATemplate      = "SomeRandomString"
+)
+
+func TestCorrectTemplate(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	err := os.Setenv(testEnvName, testEnvValue)
+	require.NoError(t, err)
+
+	domain := sandbox.NewBuilder(t).
+		SetNodesCount(0).
+		SetRegistryProxySupplier(nil).
+		SetNSMgrProxySupplier(nil).
+		SetContext(ctx).
+		Build()
+
+	// start grpc client connection and register it
+	cc, err := grpc.DialContext(ctx, grpcutils.URLToTarget(domain.Registry.URL), sandbox.DefaultDialOptions(sandbox.GenerateTestToken)...)
+	require.NoError(t, err)
+	defer func() {
+		_ = cc.Close()
+	}()
+
+	nsrc := chain.NewNetworkServiceRegistryClient(
+		nslabelsmatching.NewNetworkServiceRegistryClient(),
+		registry.NewNetworkServiceRegistryClient(cc),
+	)
+
+	ns := &registry.NetworkService{
+		Matches: []*registry.Match{
+			{
+				Routes: []*registry.Destination{
+					{
+						DestinationSelector: map[string]string{
+							destinationTestKey: destinationTestCorrectTemplate,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	ns, err = nsrc.Register(ctx, ns)
+	require.NoError(t, err)
+
+	nsrfc, err := nsrc.Find(ctx, &registry.NetworkServiceQuery{NetworkService: ns})
+	require.NoError(t, err)
+
+	ns, err = nsrfc.Recv()
+	require.NoError(t, err)
+	require.Equal(t, testEnvValue, ns.Matches[0].Routes[0].DestinationSelector[destinationTestKey])
+
+	require.NoError(t, ctx.Err())
+}
+
+func TestIncorrectTemplate(t *testing.T) {
```

Comment:
Are several tests really needed here?
You are not changing the state anywhere so you don't need the repeated test setup code before each check.
I feel like you would loose nothing if you would make it one test that just have 3 checks for 3 cases of these tests. If something failed, then you would still know which exact check failed because you have code line number in error message.

Or maybe it could be transformed into table-driven tests.

Reply:
Here is an example for the table driven tests - https://github.com/networkservicemesh/sdk/blob/main/pkg/tools/matchutils/utils_test.go.